clippy_lints/src/use_self.rs

```diff
@@ -318,3 +323,37 @@ fn lint_path_to_variant(cx: &LateContext<'_>, path: &Path<'_>) {
         span_lint(cx, span);
     }
 }
+
+/// Returns `true` if types `a` and `b` have the same lifetime parameters, otherwise returns
+/// `false`.
+///
+/// This function does not check that types `a` and `b` are the same types.
+fn same_lifetimes<'tcx>(a: MiddleTy<'tcx>, b: MiddleTy<'tcx>) -> bool {
+    use rustc_middle::ty::{Adt, GenericArgKind};
+    match (&a.kind(), &b.kind()) {
+        (&Adt(_, args_a), &Adt(_, args_b)) => {
+            args_a
+                .iter()
+                .zip(args_b.iter())
+                .all(|(arg_a, arg_b)| match (arg_a.unpack(), arg_b.unpack()) {
+                    // TODO: Handle inferred lifetimes
+                    (GenericArgKind::Lifetime(inner_a), GenericArgKind::Lifetime(inner_b)) => inner_a == inner_b,
+                    (GenericArgKind::Type(type_a), GenericArgKind::Type(type_b)) => same_lifetimes(type_a, type_b),
+                    _ => true,
```

Comment:
What would happen if `arg_a` is a Type but `arg_b` is a Lifetime? Returning `true` for those cases would not be wise.

Reply:
Can this ever happen? This function is called only when the type of `a` and `b` are the same types with the same `Const` generic args (but this doesn't check lifetime).

For `arg_a` and `arg_b` to differ in kind, this means we could have something like:
```rs
struct Foo<'a, 'b, T>{}
let x = Foo<'x, T>::new();
let y = Foo<'x, 'y, T>::new();
```
Where both `x` and `y` would co-exist, being 2 instances of `Foo` where generic arguments differ. The `_ => true` was there to ignore checks for `GenericArgKind::Const`, which are already checked prior to calling `same_lifetimes`.

My judgement might very well be clouded, but it seems impossible to me for `arg_a` and `arg_b` to have different kinds. Maybe I can count on you to _light up_ my path to spiritual elevation 💡

Reply:
I think it's possible (or maybe I misunderstood the original case):

```rust
struct Foo<'a, 'b, T>(&'a &'b T);
impl<'a, 'b, T> Foo<'a, 'b, T> {
    fn new(v: &'a &'b T) -> Self {
        Self(v)
    }
}

let x: Foo<'static, 'static, u32> = Foo::new(&&0);
let y: Foo<u32> = Foo::new(&&0);
```

Reply:
What does this code gives?

Reply:
Thinking about it some more, I don't think your case can ever happen @blyxyas, as all generics should be "computed" at this point (unless I missed something?).

Reply:
Yeah, I'm a goofy goober. After some testing, turns out that it isn't really possible :smile_cat: